From 56bf5f46112cf494090410492987f438ae000fd4 Mon Sep 17 00:00:00 2001
From: Brian Ma <chma0@nuvoton.com>
Date: Fri, 25 Mar 2022 16:42:57 +0800
Subject: [PATCH] Handle now allow execption in account service

Return message resourceCannotBeDeleted when get NowAllowed execption
during delete user instead of resourceNotFound.

Signed-off-by: Brian Ma <chma0@nuvoton.com>
---
 redfish-core/lib/account_service.hpp | 35 ++++++++++++++++++----------
 1 file changed, 23 insertions(+), 12 deletions(-)

diff --git a/redfish-core/lib/account_service.hpp b/redfish-core/lib/account_service.hpp
index 232f51c17..f2b75ae25 100644
--- a/redfish-core/lib/account_service.hpp
+++ b/redfish-core/lib/account_service.hpp
@@ -1933,22 +1933,33 @@ inline void requestAccountServiceRoutes(App& app)
                 tempObjPath /= username;
                 const std::string userPath(tempObjPath);
 
-                crow::connections::systemBus->async_method_call(
-                    [asyncResp, username](const boost::system::error_code ec) {
-                        if (ec)
+            crow::connections::systemBus->async_method_call(
+                [asyncResp, username](const boost::system::error_code ec,
+                                      const sdbusplus::message::message& msg) {
+                    if (ec)
+                    {
+                        const sd_bus_error* dbusErr = msg.get_error();
+                        if (dbusErr != nullptr &&
+                            strcmp(
+                                dbusErr->name,
+                                "xyz.openbmc_project.Common.Error.NotAllowed") ==
+                                0)
                         {
-                            messages::resourceNotFound(
-                                asyncResp->res,
-                                "#ManagerAccount.v1_4_0.ManagerAccount",
-                                username);
+                            messages::resourceCannotBeDeleted(asyncResp->res);
                             return;
                         }
 
-                        messages::accountRemoved(asyncResp->res);
-                    },
-                    "xyz.openbmc_project.User.Manager", userPath,
-                    "xyz.openbmc_project.Object.Delete", "Delete");
-            });
+                        messages::resourceNotFound(
+                            asyncResp->res,
+                            "#ManagerAccount.v1_4_0.ManagerAccount", username);
+                        return;
+                    }
+
+                    messages::accountRemoved(asyncResp->res);
+                },
+                "xyz.openbmc_project.User.Manager", userPath,
+                "xyz.openbmc_project.Object.Delete", "Delete");
+        });
 }
 
 } // namespace redfish
-- 
2.17.1

